### PR TITLE
Add ECS support

### LINKED ISSUE #39: Broken host mapping for ECS
With Elasticsearch 7.0 [elastic common schema](https://www.elastic.co/blog/introducing-the-elastic-common-schema) (ECS) was introduced. This maps the hostname to `host.name` instead of `host`. Currently `logstash_async` fails with:

> [2019-04-29T08:04:22,562][WARN ][logstash.outputs.elasticsearch] Could not index event to Elasticsearch. {:status=>400, :action=>["index", {:_id=>nil, :_index=>"logstash", :_type=>"_doc", :routing=>nil}, #<LogStash::Event:0x27275849>], :response=>{"index"=>{"_index"=>"logstash-2019.04.26-000001", "_type"=>"_doc", "_id"=>"tRAfaGoB1XS_Z1QomfCX", "status"=>400, "error"=>{"type"=>"mapper_parsing_exception", "reason"=>"object mapping for [host] tried to parse field [host] as object, but found a concrete value"}}}}

As a workaround we changed `formatter.LogstashFormatter.format` to:

```python
message = {
       	    '@timestamp': self._format_timestamp(record.created),
            '@version': '1',
       	    'host.name': self._host,  # <-- FIXED: was: 'host': ...
       	    'level': record.levelname,
            'logsource': self._logsource,
            'message': record.getMessage(),
            'pid': record.process,
            'program': self._program_name,
       	    'type': self._message_type,
        }
```

### PR DESCRIPTION
**Pull request by andriilahuta**
This is a follow up to #90.
I'm not sure if that's the right approach for it to be merged into mainstream, so I'd appreciate your input @eht16.
Fixes #39.

### CODE REVIEW

File: logstash_async/formatter.py
Comment (eht16) [inline logstash_async/formatter.py:24]: Only cosmetic: we could move this into the `LogstashFormatter` class as it is done with the others.

File: logstash_async/formatter.py
Comment (eht16) [inline logstash_async/formatter.py:267]: I wonder if it would be better to define this list explicitly in the constants module as the field names differ a lot from the legacy Logstash field list and so it might be worth to tell the user explicitly instead of just concatenating the legacy and ECS field name lists here.

File: logstash_async/formatter.py
Comment (eht16) [inline logstash_async/formatter.py:307]: Oops, never noticed this typo.

Still, even technically not necessary, I prefer to define instance variables all in `__init__()`, so it would be nice if we change this to 
```
    self._flask_version = None
```

**Review by eht16 — CHANGES_REQUESTED** [review]
Wow, looks great.

And :100: for the extensive tests.

I tested the code with the legacy formatter and it works fine.

With the ECS formatter, the final Logstash event has JSON fields like "log.syslog.hostname", will this be stored as object structure in ES or as is?

File: logstash_async/formatter.py
Comment (andriilahuta) [reply logstash_async/formatter.py:307]: Ah, hadn't occurred to me that it was meant to be `self._flask_version`. Done.

File: logstash_async/formatter.py
Comment (andriilahuta) [reply logstash_async/formatter.py]: Done.

File: logstash_async/formatter.py
Comment (andriilahuta) [reply logstash_async/formatter.py]: Makes sense.  
Maybe it would be even better to make all dotted fields top-level by default (with ability to override in constants)?

**Comment by andriilahuta** [comment]
>With the ECS formatter, the final Logstash event has JSON fields like "log.syslog.hostname", will this be stored as object structure in ES or as is?

It seems to get mapped as an object structure.
Although it is apparently [recommended](https://discuss.elastic.co/t/confusion-dot-notation-vs-sub-objects/286199) to send nested objects instead.
So, I think it probably makes sense to dedot those keys similarly to how it's done in their formatter implementation: 
https://github.com/elastic/ecs-logging-python/blob/e05b849d8493d36078543dddc9688415a74c19bf/ecs_logging/_stdlib.py#L229C26-L229C26
WDYT?

File: logstash_async/formatter.py
Comment (eht16) [reply logstash_async/formatter.py]: I'm not sure I understand what you mean exactly especially with regard to the ability to override.

What I meant in the comment above was to add a new constant like `FORMATTER_LOGSTASH_ECS_MESSAGE_FIELD_LIST` and specify the field list used by the ECS formatter as those fields are quite different from the legacy fields.
This would make it easier for users to modify the list of fields included in the event when using the ECS formatter.

Comment (andriilahuta) [reply logstash_async/formatter.py]: Isn't `*_FIELD_LIST` meant to define which fields should be top-level, not which fields to include?

Anyway, I agree that it's a good idea to define `*_FIELD_LIST` (not sure about `*_SKIP_LIST`, not many changes there) for ECS.
My idea of defining the default `FORMATTER_LOGSTASH_ECS_MESSAGE_FIELD_LIST` would look like this:
```py
ALL_DOTTED_FIELDS = object()
FORMATTER_LOGSTASH_ECS_MESSAGE_FIELD_LIST = ['@timestamp', '@version', ALL_DOTTED_FIELDS]
...
field_skip_set = self.formatter_logstash_message_field_set | {self._extra_prefix}
for key in list(message):
    if key in field_skip_set:
        continue
    if '.' in key and ALL_DOTTED_FIELDS in field_skip_set:
        continue
    message[self._extra_prefix][key] = message.pop(key)
```

**Comment by andriilahuta** [comment]
We really only need `merge_dicts` and `de_dot` functions from there which are generic enough and I had a slightly different implementation for them in mind anyway.

**Comment by eht16** [comment]
Alright, then go ahead.

File: logstash_async/formatter.py
Comment (eht16) [reply logstash_async/formatter.py]: > Isn't `*_FIELD_LIST` meant to define which fields should be top-level, not which fields to include?

`FORMATTER_LOGSTASH_MESSAGE_FIELD_LIST` is the list of field names to be included in the Logstash event.
`FORMATTER_RECORD_FIELD_SKIP_LIST` is the list of field names of the *Python* log record which should *not* be included in the Logstash event. So this one doesn't need any changes for ECS support.

> Anyway, I agree that it's a good idea to define `*_FIELD_LIST` (not sure about `*_SKIP_LIST`, not many changes there) for ECS. My idea of defining the default `FORMATTER_LOGSTASH_ECS_MESSAGE_FIELD_LIST` would look like this:
> 
> ```python
> ALL_DOTTED_FIELDS = object()
> FORMATTER_LOGSTASH_ECS_MESSAGE_FIELD_LIST = ['@timestamp', '@version', ALL_DOTTED_FIELDS]
> ...
> field_skip_set = self.formatter_logstash_message_field_set | {self._extra_prefix}
> for key in list(message):
>     if key in field_skip_set:
>         continue
>     if '.' in key and ALL_DOTTED_FIELDS in field_skip_set:
>         continue
>     message[self._extra_prefix][key] = message.pop(key)
> ```

Why do you want to generate the list?
Can't we define it manually like `FORMATTER_LOGSTASH_MESSAGE_FIELD_LIST`, e.g.:
```
    FORMATTER_LOGSTASH_ECS_MESSAGE_FIELD_LIST = [
        '@timestamp', '@version', 'host.hostname', 'log.level', 'log.syslog.hostname', ...]
```
?

Comment (andriilahuta) [reply logstash_async/formatter.py]: Looking at the implementation, `FORMATTER_LOGSTASH_MESSAGE_FIELD_LIST` seems to only be used to decide whether to put a field into `extra` or top-level, not to include/exclude it altogether:
https://github.com/eht16/python-logstash-async/blob/c60580b69103a945573586f3ee9bf09d16778f0b/logstash_async/formatter.py#L192-L205

<br>  

>Why do you want to generate the list?  

There's a lot of fields in ECS spec, most of them (all of them?) are dotted, so it looks to me like a sensible default. 
I don't see any immediate disadvantages to this, and it may even be slightly better for performance because list/set is smaller.
That said, I don't have a very strong opinion on this, we can do it manually. But it's probably going to be a long list, especially if we include fields that are not used by the lib, but may be used in user subclasses.

Comment (eht16) [reply logstash_async/formatter.py]: Oops, you are completely right.
Sorry I should have checked the code before, in my memories the `FORMATTER_LOGSTASH_MESSAGE_FIELD_LIST` was meant to remove missing fields but it is actually not.

So, since `FORMATTER_LOGSTASH_MESSAGE_FIELD_LIST` defines only the location of the fields in the Logstash event, this makes not much sense for ECS as the structure is pretty much defined by ECS, making it less useful to redefine by the user.

Comment (andriilahuta) [reply logstash_async/formatter.py]: Probably so. 
We'd need to think how to differentiate extras then, since with current implementation `_get_extra_fields` could return both top-level ecs fields and true extra fields which should/could be under the prefix.

Comment (andriilahuta) [reply logstash_async/formatter.py]: So, I've added `FORMATTER_LOGSTASH_ECS_MESSAGE_FIELD_LIST` which contains only ECS base fields (non dotted ones) and dotted fields are derived from formatter's schema.
I think we can keep `FORMATTER_LOGSTASH_ECS_MESSAGE_FIELD_LIST` in constants just in case a user wants to do something weird.

**Comment by eht16** [comment]
The changes look good and work as expected.
Great job!

There are some linting issues, after those get fixed, I think we are ready to merge.
```
logstash_async/constants.py:47:101: E501 line too long (116 > 100 characters)
logstash_async/formatter.py:268:101: E501 line too long (105 > 100 characters)
logstash_async/formatter.py:450:101: E501 line too long (103 > 100 characters)
tests/formatter_test.py:16:101: E501 line too long (106 > 100 characters)
```

**Comment by andriilahuta** [comment]
Linting issues should be fixed now.

**Comment by eht16** [comment]
I fixed the PyLint warnings:
```
Command line or configuration file:1: [E0013(bad-plugin-value), ] Plugin 'pylint.extensions.comparetozero' is impossible to load, is it installed ? ('No module named 'pylint.extensions.comparetozero'')
Command line or configuration file:1: [E0013(bad-plugin-value), ] Plugin 'pylint.extensions.emptystring' is impossible to load, is it installed ? ('No module named 'pylint.extensions.emptystring'')
```
in master.
Feel free to rebase your changes onto it or just ignore them.

The Flask warning about the version can be easily fixed:
```python
import importlib.metadata
...
importlib.metadata.version("flask")
````
if you like to add it, otherwise I will do it after this one is merged.


But the tests seem to have various problems on older Python versions and other problems on recent Python versions :(.
Do you know you can easily run them locally with `tox`?


And I don't know why I have to approve every single CI run after your pushes. Basically it should run automatically after approved once, I thought.

**Review by eht16 — APPROVED** [review]

**Comment by andriilahuta** [comment]
Tests should work now.
<br>

> And I don't know why I have to approve every single CI run after your pushes. Basically it should run automatically after approved once, I thought.

I thought so as well.
Docs mention this (https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks):
> Workflows on pull requests to public repositories from some outside contributors will not run automatically, and might need to be approved first. By default, all first-time contributors require approval to run workflows.  

The wording is a bit vague regarding subsequent approvals though.

**Comment by eht16** [comment]
Nice that you added Python 3.12 to the tests!

Thank you for your efforts and the great work!

**Comment by eht16** [comment]
I will add some docs and make a release next week.

**Pull request merged**
